# 1d Convolutional Neural Networks

In [0]:
# Imports
import numpy as np
import os
# from utils.utilities import *
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import tensorflow as tf
import pandas as pd
import tensorflow as tf
from tensorflow import keras

### About the Data

Data is from people doing one of these activities (WALKING, WALKING_UPSTAIRS, WALKING_DOWNSTAIRS, SITTING, STANDING, LAYING) <br><br>
The data is from a smartphone and provides the body acceleration, general acceleration and a gyroscope reading <br><br>
Data is from: https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones 

### Read in data from functions provided below

In [0]:
X_train, labels_train, list_ch_train = read_data_train() # train
X_test, labels_test, list_ch_test = read_data_test()

In [7]:
print(X_train.shape)

(7352, 128, 9)



#### Explaining the shape of the data 
There are 7,352 data points that are classified as doing one of those activities <br>
There are 128 time steps <br>
There are 9 values for each time step for each data point for the x,y,z values of the body 
acceleration, general acceleration and gyrscope reading. <br>

### Use train_test_split to create training and validation data and labels

In [0]:
X_train, X_val, lab_train, lab_val = train_test_split(X_train, labels_train, train_size=0.75, random_state=0)

One hot coding is when each row represents one label and it the nonzero column represents what the label is<br>
 [0,0,0,0,0,1] <br>
           ^ This label is LAYING, which was previously represented with a 5
           <br><br>

### One hot encoding functions are provided at the bottom of this notebook <br> Use these functions to convert the training, validation and testing labels

In [0]:
y_train = one_hot(lab_train)
y_val = one_hot(lab_val)
y_test = one_hot(labels_test)

### Define your Keras model <br>
You wlil want to use

See for examples on creating Keras modelshttps://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

In [17]:
# fill in with your code below
model = keras.models.Sequential()
model.add(keras.layers.Convolution1D(filters=18, kernel_size=3, strides = 1, activation = "relu", input_shape=X_train[0].shape))
model.add(keras.layers.GlobalMaxPooling1D()) # or keras.layers.MaxPooling1D(pool_size=2, padding="valid")
model.add(keras.layers.Dense(6, activation="sigmoid")) 
print(model.summary())  # view model

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_1 (Conv1D)            (None, 126, 18)           504       
_________________________________________________________________
global_max_pooling1d_1 (Glob (None, 18)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 6)                 114       
Total params: 618
Trainable params: 618
Non-trainable params: 0
_________________________________________________________________
None


### Compile your model

In [0]:
lr =  0.0008
model.compile( optimizer=tf.train.AdamOptimizer(lr), loss=keras.losses.categorical_crossentropy, metrics = ['accuracy'] )

### Fit your model

In [20]:
history = model.fit( X_train,
                    y_train,
                    epochs= 100,
                    batch_size= 600,
                    validation_data=(X_val, y_val),
                    verbose=1 
                    )

Train on 5514 samples, validate on 1838 samples
Epoch 1/100
5514/5514 [==============================] - 1s 203us/step - loss: 1.7821 - acc: 0.1884 - val_loss: 1.7635 - val_acc: 0.1986
Epoch 2/100
5514/5514 [==============================] - 1s 203us/step - loss: 1.7513 - acc: 0.2053 - val_loss: 1.7360 - val_acc: 0.2307
Epoch 3/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.7242 - acc: 0.2441 - val_loss: 1.7111 - val_acc: 0.2715
Epoch 4/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.6988 - acc: 0.2907 - val_loss: 1.6875 - val_acc: 0.3112
Epoch 5/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.6746 - acc: 0.3322 - val_loss: 1.6640 - val_acc: 0.3542
Epoch 6/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.6503 - acc: 0.3576 - val_loss: 1.6403 - val_acc: 0.3678
Epoch 7/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.6258 - acc: 0.3743 - val_loss: 1.6164 - val_acc: 

Epoch 12/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.5063 - acc: 0.5044 - val_loss: 1.5016 - val_acc: 0.5256
Epoch 13/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.4838 - acc: 0.5163 - val_loss: 1.4798 - val_acc: 0.5365
Epoch 14/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.4617 - acc: 0.5263 - val_loss: 1.4582 - val_acc: 0.5403
Epoch 15/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.4399 - acc: 0.5357 - val_loss: 1.4367 - val_acc: 0.5490
Epoch 16/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.4180 - acc: 0.5361 - val_loss: 1.4149 - val_acc: 0.5484
Epoch 17/100
5514/5514 [==============================] - 1s 224us/step - loss: 1.3963 - acc: 0.5433 - val_loss: 1.3932 - val_acc: 0.5517
Epoch 18/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.3745 - acc: 0.5481 - val_loss: 1.3716 - val_acc: 0.5550
Epoch 19/100
5514/5514 [==========

5514/5514 [==============================] - 1s 224us/step - loss: 1.2662 - acc: 0.5713 - val_loss: 1.2617 - val_acc: 0.5773
Epoch 24/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.2434 - acc: 0.5753 - val_loss: 1.2382 - val_acc: 0.5827
Epoch 25/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.2196 - acc: 0.5811 - val_loss: 1.2134 - val_acc: 0.5936
Epoch 26/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.1949 - acc: 0.5889 - val_loss: 1.1874 - val_acc: 0.5996
Epoch 27/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.1688 - acc: 0.5974 - val_loss: 1.1600 - val_acc: 0.6088
Epoch 28/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.1415 - acc: 0.6085 - val_loss: 1.1315 - val_acc: 0.6251
Epoch 29/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.1128 - acc: 0.6184 - val_loss: 1.1018 - val_acc: 0.6349
Epoch 30/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.9450 - acc: 0.7466 - val_loss: 0.9302 - val_acc: 0.7682
Epoch 35/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.9088 - acc: 0.7642 - val_loss: 0.8945 - val_acc: 0.7867
Epoch 36/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.8734 - acc: 0.7838 - val_loss: 0.8599 - val_acc: 0.8041
Epoch 37/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.8392 - acc: 0.7967 - val_loss: 0.8262 - val_acc: 0.8128
Epoch 38/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.8060 - acc: 0.8081 - val_loss: 0.7942 - val_acc: 0.8215
Epoch 39/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.7743 - acc: 0.8185 - val_loss: 0.7631 - val_acc: 0.8264
Epoch 40/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.7437 - acc: 0.8273 - val_loss: 0.7333 - val_acc: 0.8335
Epoch 41/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.6077 - acc: 0.8636 - val_loss: 0.6019 - val_acc: 0.8624
Epoch 46/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.5840 - acc: 0.8698 - val_loss: 0.5790 - val_acc: 0.8667
Epoch 47/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.5616 - acc: 0.8752 - val_loss: 0.5580 - val_acc: 0.8765
Epoch 48/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.5403 - acc: 0.8796 - val_loss: 0.5375 - val_acc: 0.8798
Epoch 49/100
5514/5514 [==============================] - 1s 227us/step - loss: 0.5202 - acc: 0.8825 - val_loss: 0.5185 - val_acc: 0.8825
Epoch 50/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.5013 - acc: 0.8845 - val_loss: 0.5003 - val_acc: 0.8841
Epoch 51/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.4835 - acc: 0.8886 - val_loss: 0.4833 - val_acc: 0.8863
Epoch 52/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.4102 - acc: 0.8968 - val_loss: 0.4136 - val_acc: 0.8955
Epoch 57/100
5514/5514 [==============================] - 1s 222us/step - loss: 0.3986 - acc: 0.8983 - val_loss: 0.4021 - val_acc: 0.8961
Epoch 58/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.3871 - acc: 0.9003 - val_loss: 0.3912 - val_acc: 0.8955
Epoch 59/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.3768 - acc: 0.8999 - val_loss: 0.3813 - val_acc: 0.8983
Epoch 60/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.3667 - acc: 0.9032 - val_loss: 0.3719 - val_acc: 0.9015
Epoch 61/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.3576 - acc: 0.9071 - val_loss: 0.3628 - val_acc: 0.9059
Epoch 62/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.3487 - acc: 0.9106 - val_loss: 0.3543 - val_acc: 0.9064
Epoch 63/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.3119 - acc: 0.9182 - val_loss: 0.3181 - val_acc: 0.9157
Epoch 68/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.3056 - acc: 0.9175 - val_loss: 0.3120 - val_acc: 0.9162
Epoch 69/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2998 - acc: 0.9180 - val_loss: 0.3062 - val_acc: 0.9146
Epoch 70/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.2940 - acc: 0.9204 - val_loss: 0.3009 - val_acc: 0.9189
Epoch 71/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.2887 - acc: 0.9209 - val_loss: 0.2958 - val_acc: 0.9178
Epoch 72/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.2837 - acc: 0.9213 - val_loss: 0.2908 - val_acc: 0.9189
Epoch 73/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2790 - acc: 0.9213 - val_loss: 0.2859 - val_acc: 0.9178
Epoch 74/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.2579 - acc: 0.9251 - val_loss: 0.2659 - val_acc: 0.9200
Epoch 79/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.2546 - acc: 0.9242 - val_loss: 0.2624 - val_acc: 0.9195
Epoch 80/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.2510 - acc: 0.9251 - val_loss: 0.2591 - val_acc: 0.9206
Epoch 81/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.2475 - acc: 0.9262 - val_loss: 0.2564 - val_acc: 0.9211
Epoch 82/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2445 - acc: 0.9275 - val_loss: 0.2527 - val_acc: 0.9206
Epoch 83/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2415 - acc: 0.9264 - val_loss: 0.2497 - val_acc: 0.9206
Epoch 84/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.2385 - acc: 0.9278 - val_loss: 0.2468 - val_acc: 0.9211
Epoch 85/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 223us/step - loss: 0.2253 - acc: 0.9295 - val_loss: 0.2340 - val_acc: 0.9233
Epoch 90/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.2231 - acc: 0.9291 - val_loss: 0.2317 - val_acc: 0.9227
Epoch 91/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2207 - acc: 0.9298 - val_loss: 0.2296 - val_acc: 0.9227
Epoch 92/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.2186 - acc: 0.9311 - val_loss: 0.2276 - val_acc: 0.9227
Epoch 93/100
5514/5514 [==============================] - 1s 222us/step - loss: 0.2165 - acc: 0.9305 - val_loss: 0.2253 - val_acc: 0.9233
Epoch 94/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2144 - acc: 0.9305 - val_loss: 0.2233 - val_acc: 0.9233
Epoch 95/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2125 - acc: 0.9302 - val_loss: 0.2214 - val_acc: 0.9244
Epoch 96/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.2034 - acc: 0.9320 - val_loss: 0.2127 - val_acc: 0.9238


### Code below prints the test loss and accuracy <br>
### Before you print your accuracy and loss, play around with the parameters!!

In [21]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

('Test loss:', 0.3620428957189185)
('Test accuracy:', 0.8686800135731252)


### Provided Functions Below

In [0]:
def read_data_train():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_train.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def read_data_test():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_test.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def one_hot(labels, n_class = 6):
	""" One-hot encoding """
	expansion = np.eye(n_class)
	y = expansion[:, labels-1].T
	assert y.shape[1] == n_class, "Wrong number of labels!"

	return y

In [0]:
def get_batches(X, y, batch_size = 100):
	""" Return a generator for batches """
	n_batches = len(X) // batch_size
	X, y = X[:n_batches*batch_size], y[:n_batches*batch_size]

	# Loop over batches and yield
	for b in range(0, len(X), batch_size):
		yield X[b:b+batch_size], y[b:b+batch_size]